Gems/AWSGameLift/Code/AWSGameLiftClient/Source/AWSGameLiftClientLocalTicketTracker.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Interface/Interface.h>
+#include <AzCore/std/bind/bind.h>
+#include <AzCore/std/smart_ptr/shared_ptr.h>
+#include <AzFramework/Session/ISessionHandlingRequests.h>
+
+#include <AWSGameLiftClientLocalTicketTracker.h>
+#include <AWSGameLiftSessionConstants.h>
+#include <Request/IAWSGameLiftInternalRequests.h>
+
+#include <aws/core/utils/Outcome.h>
+#include <aws/gamelift/GameLiftClient.h>
+#include <aws/gamelift/model/DescribeMatchmakingRequest.h>
+
+namespace AWSGameLift
+{
+    AWSGameLiftClientLocalTicketTracker::AWSGameLiftClientLocalTicketTracker()
+        : m_status(TicketTrackerStatus::Idle)
+        , m_pollingPeriodInMS(AWSGameLiftClientDefaultPollingPeriodInMS)
+    {
+    }
+
+    void AWSGameLiftClientLocalTicketTracker::ActivateTracker()
+    {
+        AZ::Interface<IAWSGameLiftMatchmakingTicketRequests>::Register(this);
+    }
+
+    void AWSGameLiftClientLocalTicketTracker::DeactivateTracker()
+    {
+        AZ::Interface<IAWSGameLiftMatchmakingTicketRequests>::Unregister(this);
+        StopPolling();
+    }
+
+    void AWSGameLiftClientLocalTicketTracker::StartPolling(
+        const AZStd::string& ticketId, const AZStd::string& playerId)
+    {
+        AZStd::lock_guard<AZStd::mutex> lock(m_trackerMutex);
+        if (m_status == TicketTrackerStatus::Running)
+        {
+            AZ_TracePrintf(AWSGameLiftClientLocalTicketTrackerName, "Matchmaking ticket tracker is running.");
+            return;
+        }
+        m_status = TicketTrackerStatus::Running;
+        m_trackerThread = AZStd::thread(AZStd::bind(
+            &AWSGameLiftClientLocalTicketTracker::ProcessPolling, this, ticketId, playerId));
+    }
+
+    void AWSGameLiftClientLocalTicketTracker::StopPolling()
+    {
+        AZStd::lock_guard<AZStd::mutex> lock(m_trackerMutex);
+        m_status = TicketTrackerStatus::Idle;
+        if (m_trackerThread.joinable())
+        {
+            m_trackerThread.join();
+        }
+    }
+
+    void AWSGameLiftClientLocalTicketTracker::ProcessPolling(
+        const AZStd::string& ticketId, const AZStd::string& playerId)
+    {
+        while (m_status == TicketTrackerStatus::Running)
+        {
+            auto gameliftClient = AZ::Interface<IAWSGameLiftInternalRequests>::Get()->GetGameLiftClient();
```

Comment:
nit: why are you getting the client every time in the while loop? should this be outside of the loop?

Reply:
safety check for incorrect use case, which developer may destruct client but leave ticket tracker running on background